docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,228 @@
+    CurrentCPUUtilizationPercentage *int32 `json:"currentCPUUtilizationPercentage,omitempty"`
+
+    // the last read state of the metrics used by this autoscaler
+    CurrentMetrics []MetricStatus `json:"currentMetrics" protobuf:"bytes,5,rep,name=currentMetrics"`
+}
+
+// the status of a single metric
+type MetricStatus struct {
+    // metric describing a single Kubernetes object
+    Object *ObjectMetricStatus `json:"object,omitemtpy"`
+
+    // metric describing pods in the scale target
+    Pods *PodsMetricStatus `json:"pods,omitemtpy"`
+}
+
+// a metric describing a Kubernetes object
+type ObjectMetricStatus struct {
+    // the described Kubernetes object
+    Target CrossVersionObjectReference `json:"target"`
+
+    // the name of the metric in question
+    MetricName string `json:"metricName"`
+
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
+}
+
+// metric describing pods in the scale target
+type PodsMetricStatus struct {
+
+    // the name of the metric in question
+    MetricName `json:"metricName"`
+
+    // the current value of the metric (as a quantity)
+    CurrentValue resource.Quantity `json:"targetValue"`
+}
+```
+
+### Example ###
+
+In this example, we scale based on the `hits-per-second` value recorded as
+describing a service in our namespace, plus the CPU usage of the pods in
+the ReplicationController being autoscaled.
+
+```yaml
+kind: HorizontalPodAutoscaler
+apiVersion: autoscaling/v2alpha1
+spec:
+  scaleTargetRef:
+    kind: ReplicationController
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  targetCPUUtilizationPercentage: 80
+  metrics:
+  - object:
+      target:
+        kind: Service
+        name: Frontend
+      metricName: hits-per-second
+      targetValue: 1k
+```
+
+### Alternatives and Future Considerations ###
+
+Since the new design mirrors volume plugins (and similar APIs), it makes
+it relatively easy to introduce new fields in a backwards-compatible way:
+we simply introduce a new field in `MetricSpec` as a new "metric type".
+
+#### External ####
+
+As a potential future consideration, we may wish to include a metric
+source type of `External`, which simply has an opaque metric type and
+target value (as strings).  This would allow custom autoscalers with
+custom source mechanics to be written, while still providing the same
+indication that a controller was being autoscaled.
+
+#### Resource Percentages ####
+
+In the current design, CPU percentage metrics are a completely separate
+field from the rest of the metrics.  While this simplifies the common case
+of scaling on CPU, it makes the overall design a bit more clunky.  It also
+doesn't provide a method for scaling based on memory percentage.
+
+Two alternatives to the method proposed above are shown the object below.
+The former is more compact, but precludes scaling memory as a percentage
```

Comment:
I like having CPU separately for several reasons. First of all CPU data comes from a separate pipeline (see new monitoring architecture for monitoring) that is supposed to be:
a) fast
b) always available (it's part of core deployment)
c) work in identical way everywhere
d) report information which are consistent with information in all other places that use them (e.g. scheduler, or kubelet, or user when specifying request/limit)

For custom metrics we don't have any of those guarantees. What's more they are far less portable (they work differently with different monitoring solutions).

For those reasons I believe we should treat CPU separately. Whether it's a special field in `Metric` or a top level field is less important for me, though IMO it's more clear if we have a group of metrics which are different in some way to be grouped together (like you have in this proposal).

With regard to memory it's actually very tricky to scale based on this, as it's very hard to build a proper model for scaling based on this signal. We've been doing some research around that and decided in the end it's not worth the effort. But I'm open for suggestions here :)